### DISCUSSION TITLE
SUPER WEIRD OR SUPER SILLY? PLEASE HELP.

### DISCUSSION THREAD

**Discussion by omgoshjosh** [react-hook-form/react-hook-form#3417] [Q&A] [answered]
Hello,

I am still trying to reproduce a bug i'm seeing in a CSB, but it may just be bad or conflicting config or usage of this (awesome) library on my part. Since I haven't been able to repro it in a CSB, I decided to open a discussion and I'm hoping I'm just hitting some gotcha, or this is just my misusage of RHF. Problem is described below.

When appending a new item to useFieldArray, my values are getting wiped out (to the default value empty string) in the the process.

I am building a modal that has 2 main states, in the default state, we are inputting a bunch of emails into useFieldArray objects like

```javascript
{ email: 'theEmail@goes.here' }
```

and in the other state, we are turning those emails into a comma separated string like

```javascript
'theEmail@goes.here, and@also.here'
```

My question is based on what happens when returning to the default state with shouldUnregister false.

I'm mapping the fields to a controlled custom input

```jsx
const { theArray, theString } = watch(['theArray', 'theString']);

showTextArea ? <textarea value={theString}>{/*this textarea is fine*/}
: fields.map(({ id }, index) => (
      <div key={id}>
          <Controller render={({ onChange, onBlur,  }) => (
              {/* the email below is getting wiped out */}
              <CustomInput initialValue={theArray[index].email} onChange={onChange} onBlur={onBlur}/>
          )} />
      </div>
  );
```

I'm splitting the comma separated string and calling `setValue()` like so

```javascript
const emails = theString.split(',');
emails.forEach((email, i) => setValue(`theArray[${i}].email`, email) );
// or alternatively with the whole object
// emails.forEach((email, i) => setValue(`theArray[${i}]`, { email }) );
```

Once in this state, with an input populated from the call to setValue, at some point RHF calls `getFieldsValues()` and I am trying to find out why `getFieldsValues()` sends in undefined when calling `getFieldValue()`?

https://github.com/react-hook-form/react-hook-form/blob/master/src/logic/getFieldsValues.ts#L30

here's the snippet in question
```typescript
      output[name as InternalFieldName<TFieldValues>] = getFieldValue(
        fieldsRef,
        name,
        undefined,
        excludeDisabled,
      );
```

I am also looking at this return statement as another possible avenue to investigate

```javascript
  return deepMerge(
    transformToNestObject({ ...((shallowFieldsStateRef || {}).current || {}) }),
    transformToNestObject(output),
  )
```

**Comment by bluebill1049** [comment]
hmmm did you set up `defaultValue` for `Controller`?

**Reply by omgoshjosh** [reply]
I did do that but with `reset()` like so

```javascript
const { reset } = useForm({ mode: 'onChange', shouldUnregister: false, });
useEffect(() => {
    if(showModal) {
        reset({
            theString: '',
            theArray: getInitialArray()
        });
    } else {
        remove();
    }
}, [showModal]);

```

[which according to docs, should be ok](url)

> When invoking reset({ value }) without supply defaultValues at useForm, hook form will replace defaultValues with value object which you have supplied.

PS: i am still working on the CSB and will see if I can repro it, thanks for the eyeballs.

**Reply by bluebill1049** [reply]
no worries, mate. i wonder what's inside `getInitialArray`? no rush tho, I am subscribed to this question, will wait for your CSB.

**Reply by omgoshjosh** [reply]
```javascript
// this is simplified, i am actually putting other data on here
// they would be level which is a boolean prop passed into the
// modal, and toggleData which is shaped like { [key: string]: boolean }
const INITIAL_DATA = [{ email: '' }, { email: '' }];

function getInitialArray() {
    const result = INITIAL_DATA.map(initialObj => ({ email: '' }));
    return result;
}
```

**Comment by omgoshjosh** [comment, marked as answer]
I think what is working for me right now is just to get rid of the controller and use the register function (i do have control of the custom input)

```jsx
function CustomInput(props, ref) {
    return <input ref={ref} />
}

const CustomInput = React.forwardRef(_CustomInput);

<CustomInput ref={register()} />
```

still very mysterious!

**Reply by bluebill1049** [reply]
strange, i am keen to resolve the problem if you can supply a CSB, whenever you see that happened again.
